File: packages/core-ts/src/storage/machine-id.ts
import { execFile } from 'node:child_process'
import { readFile } from 'node:fs/promises'
import { hostname, networkInterfaces, userInfo } from 'node:os'
import { promisify } from 'node:util'

const execFileAsync = promisify(execFile)

let cachedMachineId: string | null = null

export function clearMachineIdCacheForTesting(): void {
  cachedMachineId = null
}

export async function getMachineId(): Promise<string> {
  if (cachedMachineId) return cachedMachineId

  try {
    if (process.platform === 'linux') {
      cachedMachineId = (await readFile('/etc/machine-id', 'utf-8')).trim()
      return cachedMachineId
    }
    if (process.platform === 'darwin') {
      const { stdout } = await execFileAsync('ioreg', ['-rd1', '-c', 'IOPlatformExpertDevice'])
      const match = stdout.match(/"IOPlatformUUID"\s*=\s*"([^"]+)"/)
      if (match) {
        cachedMachineId = match[1]
        return cachedMachineId
      }
    }
    if (process.platform === 'win32') {
      const { stdout } = await execFileAsync('reg', [
        'query',
        'HKLM\\SOFTWARE\\Microsoft\\Cryptography',
        '/v',
        'MachineGuid'
      ])
      const match = stdout.match(/MachineGuid\s+REG_SZ\s+(\S+)/)
      if (match) {
        cachedMachineId = match[1]
        return cachedMachineId
      }
    }
  } catch {
    /* fallback below */
  }

  // Fallback: hostname + first MAC address
  const nics = networkInterfaces()

  // Optimization: Manual loops are significantly faster than .flat().find() in Bun
  // as they avoid intermediate memory allocations and functional iteration overhead.
  let mac: string | undefined
  const values = Object.values(nics)
  for (let i = 0; i < values.length; i++) {
    const ifaces = values[i]
    if (ifaces) {
      for (let j = 0; j < ifaces.length; j++) {
        const n = ifaces[j]
        if (n && !n.internal && n.mac !== '00:00:00:00:00:00') {
          mac = n.mac
          break
        }
      }
    }
    if (mac) break
  }

  cachedMachineId = `${hostname()}-${mac ?? 'unknown'}`
  return cachedMachineId
}

export function getUsername(): string {
  try {
    return userInfo().username
  } catch {
    return process.env.USER ?? process.env.USERNAME ?? 'unknown'
  }
}

File: packages/core-py/src/mcp_core/storage/machine_id.py
"""Machine ID and username detection (cross-platform)."""

import getpass
import os
import re
import socket
import subprocess
import sys
import uuid
from functools import lru_cache


@lru_cache(maxsize=1)
def get_machine_id() -> str:
    """Get a stable machine identifier.

    Platform-specific:
    - Linux: /etc/machine-id
    - macOS: IOPlatformUUID via ioreg
    - Windows: MachineGuid from registry

    Falls back to hostname + first MAC address.

    Uses ``sys.platform`` (string compare, no syscall) instead of
    ``platform.system()`` which on Windows triggers WMI via
    ``platform.win32_ver()``; degraded WMI services hang the process.

    Returns:
        Machine identifier string.
    """
    plat = sys.platform
    try:
        if plat.startswith("linux"):
            with open("/etc/machine-id") as f:
                return f.read().strip()

        if plat == "darwin":
            result = subprocess.run(
                ["ioreg", "-rd1", "-c", "IOPlatformExpertDevice"],
                capture_output=True,
                text=True,
                timeout=5,
            )
            match = re.search(r'"IOPlatformUUID"\s*=\s*"([^"]+)"', result.stdout)
            if match:
                return match.group(1)

        if plat == "win32":
            result = subprocess.run(
                [
                    "reg",
                    "query",
                    r"HKLM\SOFTWARE\Microsoft\Cryptography",
                    "/v",
                    "MachineGuid",
                ],
                capture_output=True,
                text=True,
                timeout=5,
            )
            match = re.search(r"MachineGuid\s+REG_SZ\s+(\S+)", result.stdout)
            if match:
                return match.group(1)
    except Exception:
        pass

    # Fallback: hostname + first MAC address
    mac = _get_first_mac()
    return f"{socket.gethostname()}-{mac}"


def _get_first_mac() -> str:
    """Get first non-loopback MAC address, or 'unknown'."""
    try:
        mac_int = uuid.getnode()
        # uuid.getnode() returns a random MAC if it can't find one,
        # indicated by the multicast bit being set
        if (mac_int >> 40) & 1:
            return "unknown"
        mac_str = ":".join([f"{(mac_int >> (8 * i)) & 0xFF:02x}" for i in range(5, -1, -1)])
        if mac_str == "00:00:00:00:00:00":
            return "unknown"
        return mac_str
    except Exception:
        return "unknown"


def get_username() -> str:
    """Get current OS username.

    Returns:
        Username string.
    """
    try:
        return getpass.getuser()
    except Exception:
        return os.environ.get("USER", os.environ.get("USERNAME", "unknown"))
